codec/aminojson/json_marshal.go

```diff
@@ -0,0 +1,247 @@
+package aminojson
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	cosmos_proto "github.com/cosmos/cosmos-proto"
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"cosmossdk.io/api/amino"
+)
+
+type JSONMarshaller interface {
+	MarshalAmino(proto.Message) ([]byte, error)
+}
+
+type AminoJSON struct {
+	// maps cosmos_proto.scalar -> zero value factory
+	zeroValues      map[string]func() protoreflect.Value
+	messageEncoders map[string]func(message protoreflect.Message) protoreflect.Value
+}
+
+func NewAminoJSON() AminoJSON {
+	aj := AminoJSON{
+		zeroValues: map[string]func() protoreflect.Value{
+			"cosmos.Dec": func() protoreflect.Value {
+				return protoreflect.ValueOfString("0")
+			},
+		},
+		messageEncoders: map[string]func(message protoreflect.Message) protoreflect.Value{
+			"key_field": func(message protoreflect.Message) protoreflect.Value {
+				// todo
+				// err handling
+				keyField := message.Get(message.Descriptor().Fields().ByName("key"))
+				bz := keyField.Bytes()
+				return protoreflect.ValueOfBytes(bz)
+			},
+		},
+	}
+	return aj
+}
+
+func MarshalAmino(message proto.Message) ([]byte, error) {
+	buf := &bytes.Buffer{}
+	aj := NewAminoJSON()
+	vmsg := protoreflect.ValueOfMessage(message.ProtoReflect())
+	err := aj.marshal(vmsg, buf)
+	return buf.Bytes(), err
+}
```

Comment:
Given the above comment that there are sometimes custom message encodings, I wonder if this function is the best API.

Maybe a 2-step API is more flexible:

```go
aj := NewAminoJSON()
// optionally, define custom message encodings
aj.Marshal() // or aj.MarshalJSON()
```

which is more inline with `protojson.MarshalOptions{}.Marshal` API

Reply:
Done.. let me know what you think.